File: lib/game/ludo_rules.dart
import '../models/ludo_models.dart';
import 'classic_board.dart';

/// Pure, deterministic Ludo rules shared by the UI, bots, and persistence
/// layer. This class deliberately has no Firebase or Flutter dependencies.
class LudoRules {
  const LudoRules._();

  static const int basePosition = -1;
  static const int outerTrackLastPosition = 51;
  static const int goalPosition = 5;

  static const Set<int> safeGlobalPositions = {
    3,
    8,
    11,
    16,
    21,
    24,
    29,
    34,
    37,
    42,
    47,
    50,
  };

  static bool isValidMove(LudoPiece piece, int diceValue) {
    if (diceValue < 1 || diceValue > 6) return false;
    if (piece.inHome && piece.pos == goalPosition) return false;
    if (piece.pos == basePosition) return diceValue == 6;
    if (piece.inHome) return piece.pos + diceValue <= goalPosition;
    return piece.pos >= 0 && piece.pos <= outerTrackLastPosition;
  }

  static bool hasValidMove(Iterable<LudoPiece> pieces, int diceValue) {
    return pieces.any((piece) => isValidMove(piece, diceValue));
  }

  static List<ActiveMoveStep> buildMoveSteps(LudoPiece piece, int diceValue) {
    if (!isValidMove(piece, diceValue)) return const <ActiveMoveStep>[];

    final steps = <ActiveMoveStep>[
      ActiveMoveStep(pos: piece.pos, inHome: piece.inHome),
    ];
    var remaining = piece.pos == basePosition ? 1 : diceValue;
    var position = piece.pos;
    var inHome = piece.inHome;

    while (remaining-- > 0) {
      if (position == basePosition) {
        position = 0;
        inHome = false;
      } else if (inHome) {
        position++;
      } else {
        position++;
        if (position > outerTrackLastPosition) {
          position = 0;
          inHome = true;
        }
      }
      steps.add(ActiveMoveStep(pos: position, inHome: inHome));
    }

    return steps;
  }

  static ActiveMoveStep destination(LudoPiece piece, int diceValue) {
    final steps = buildMoveSteps(piece, diceValue);
    return steps.isEmpty
        ? ActiveMoveStep(pos: piece.pos, inHome: piece.inHome)
        : steps.last;
  }

  static bool reachesGoal(LudoPiece before, ActiveMoveStep after) {
    return after.inHome &&
        after.pos == goalPosition &&
        !(before.inHome && before.pos == goalPosition);
  }

  static bool hasCompletedAllPieces(Iterable<LudoPiece> pieces) {
    return pieces.isNotEmpty &&
        pieces.every((piece) => piece.inHome && piece.pos == goalPosition);
  }

  static bool grantsExtraTurn({
    required int diceValue,
    required bool didCapture,
    required bool didReachGoal,
  }) {
    return diceValue == 6 || didCapture || didReachGoal;
  }

  static String nextActivePlayer({
    required List<String> players,
    required String currentPlayerId,
    required Iterable<String> finishedPlayers,
  }) {
    if (players.isEmpty) return '';
    final finished = finishedPlayers.toSet();
    final currentIndex = players.indexOf(currentPlayerId);
    final startIndex = currentIndex < 0 ? -1 : currentIndex;

    for (var offset = 1; offset <= players.length; offset++) {
      final candidate = players[(startIndex + offset) % players.length];
      if (!finished.contains(candidate)) return candidate;
    }
    return '';
  }

  static NoValidMoveTurn resolveNoValidMove({
    required List<String> players,
    required String currentPlayerId,
    required Iterable<String> finishedPlayers,
    required int diceValue,
  }) {
    final keepsTurn = diceValue == 6;
    return NoValidMoveTurn(
      nextPlayerId: keepsTurn
          ? currentPlayerId
          : nextActivePlayer(
              players: players,
              currentPlayerId: currentPlayerId,
              finishedPlayers: finishedPlayers,
            ),
      keepsTurn: keepsTurn,
    );
  }

  static List<String> finishOrderAfterMove({
    required List<String> players,
    required Iterable<String> currentFinishOrder,
    required String movingPlayerId,
    required bool movingPlayerFinished,
  }) {
    final result = <String>[];
    for (final playerId in currentFinishOrder) {
      if (players.contains(playerId) && !result.contains(playerId)) {
        result.add(playerId);
      }
    }
    if (movingPlayerFinished && !result.contains(movingPlayerId)) {
      result.add(movingPlayerId);
    }

    final unfinished = players.where((id) => !result.contains(id)).toList();
    if (movingPlayerFinished && unfinished.length == 1) {
      result.add(unfinished.single);
    }
    return result;
  }

  static bool isMatchFinished({
    required List<String> players,
    required List<String> finishOrder,
  }) {
    return players.length >= 2 && finishOrder.length == players.length;
  }

  static CaptureResult applyCaptures({
    required Map<String, List<LudoPiece>> pieces,
    required Map<String, int> playerSeats,
    required List<String> players,
    required String movingPlayerId,
    required ActiveMoveStep destination,
  }) {
    final updated = <String, List<LudoPiece>>{
      for (final entry in pieces.entries)
        entry.key: List<LudoPiece>.from(entry.value),
    };
    if (destination.inHome) {
      return CaptureResult(pieces: updated, didCapture: false);
    }

    final movingSeat = _seatFor(players, playerSeats, movingPlayerId);
    final globalDestination = ClassicBoard.globalPathIndexForSeat(
      movingSeat,
      destination.pos,
    );
    if (safeGlobalPositions.contains(globalDestination)) {
      return CaptureResult(pieces: updated, didCapture: false);
    }

    var didCapture = false;
    final capturedPieces = <ActiveMoveCapture>[];
    for (final opponentId in players) {
      if (opponentId == movingPlayerId) continue;
      final opponentSeat = _seatFor(players, playerSeats, opponentId);
      updated[opponentId] = (updated[opponentId] ?? const <LudoPiece>[]).map((
        piece,
      ) {
        if (piece.pos == basePosition || piece.inHome) return piece;
        final globalPosition = ClassicBoard.globalPathIndexForSeat(
          opponentSeat,
          piece.pos,
        );
        if (globalPosition != globalDestination) return piece;
        didCapture = true;
        capturedPieces.add(
          ActiveMoveCapture(
            playerId: opponentId,
            pieceId: piece.id,
            from: ActiveMoveStep(pos: piece.pos, inHome: piece.inHome),
          ),
        );
        return piece.copyWith(pos: basePosition, inHome: false);
      }).toList();
    }
    return CaptureResult(
      pieces: updated,
      didCapture: didCapture,
      capturedPieces: capturedPieces,
    );
  }

  static bool needsLegacyActionRecovery(LudoGame game) {
    return needsActionRecovery(
      activeDiceRoll: game.activeDiceRoll,
      activeMove: game.activeMove,
    );
  }

  static bool needsActionRecovery({
    ActiveDiceRoll? activeDiceRoll,
    ActiveMove? activeMove,
  }) {
    return (activeDiceRoll != null && !activeDiceRoll.stateApplied) ||
        (activeMove != null && !activeMove.stateApplied);
  }

  static int _seatFor(
    List<String> players,
    Map<String, int> playerSeats,
    String playerId,
  ) {
    return (playerSeats[playerId] ?? players.indexOf(playerId))
        .clamp(0, 3)
        .toInt();
  }
}

class CaptureResult {
  final Map<String, List<LudoPiece>> pieces;
  final bool didCapture;
  final List<ActiveMoveCapture> capturedPieces;

  const CaptureResult({
    required this.pieces,
    required this.didCapture,
    this.capturedPieces = const <ActiveMoveCapture>[],
  });
}

class NoValidMoveTurn {
  final String nextPlayerId;
  final bool keepsTurn;

  const NoValidMoveTurn({required this.nextPlayerId, required this.keepsTurn});
}

File: functions/src/ludo_rules.ts
export const BASE_POSITION = -1;
export const OUTER_TRACK_LAST_POSITION = 51;
export const GOAL_POSITION = 5;
export const SAFE_GLOBAL_POSITIONS = new Set([3, 8, 11, 16, 21, 24, 29, 34, 37, 42, 47, 50]);
export const START_OFFSETS = [11, 24, 37, 50] as const;

export interface Piece {
  id: number;
  pos: number;
  inHome: boolean;
}

export interface MoveStep {
  pos: number;
  inHome: boolean;
}

export interface MoveCapture {
  playerId: string;
  pieceId: number;
  from: MoveStep;
}

export interface CaptureResult {
  pieces: Record<string, Piece[]>;
  didCapture: boolean;
  capturedPieces: MoveCapture[];
}

export function isValidMove(piece: Piece, diceValue: number): boolean {
  if (!Number.isInteger(diceValue) || diceValue < 1 || diceValue > 6) return false;
  if (piece.inHome && piece.pos === GOAL_POSITION) return false;
  if (piece.pos === BASE_POSITION) return diceValue === 6;
  if (piece.inHome) return piece.pos + diceValue <= GOAL_POSITION;
  return piece.pos >= 0 && piece.pos <= OUTER_TRACK_LAST_POSITION;
}

export function hasValidMove(pieces: readonly Piece[], diceValue: number): boolean {
  return pieces.some((piece) => isValidMove(piece, diceValue));
}

export function buildMoveSteps(piece: Piece, diceValue: number): MoveStep[] {
  if (!isValidMove(piece, diceValue)) return [];
  const steps: MoveStep[] = [{pos: piece.pos, inHome: piece.inHome}];
  let remaining = piece.pos === BASE_POSITION ? 1 : diceValue;
  let position = piece.pos;
  let inHome = piece.inHome;
  while (remaining-- > 0) {
    if (position === BASE_POSITION) {
      position = 0;
      inHome = false;
    } else if (inHome) {
      position += 1;
    } else {
      position += 1;
      if (position > OUTER_TRACK_LAST_POSITION) {
        position = 0;
        inHome = true;
      }
    }
    steps.push({pos: position, inHome});
  }
  return steps;
}

export function destination(piece: Piece, diceValue: number): MoveStep {
  const steps = buildMoveSteps(piece, diceValue);
  return steps.length === 0 ? {pos: piece.pos, inHome: piece.inHome} : steps[steps.length - 1];
}

export function reachesGoal(before: Piece, after: MoveStep): boolean {
  return after.inHome && after.pos === GOAL_POSITION && !(before.inHome && before.pos === GOAL_POSITION);
}

export function hasCompletedAllPieces(pieces: readonly Piece[]): boolean {
  return pieces.length > 0 && pieces.every((piece) => piece.inHome && piece.pos === GOAL_POSITION);
}

export function grantsExtraTurn(diceValue: number, didCapture: boolean, didReachGoal: boolean): boolean {
  return diceValue === 6 || didCapture || didReachGoal;
}

export function nextActivePlayer(
  players: readonly string[],
  currentPlayerId: string,
  finishedPlayers: readonly string[],
): string {
  if (players.length === 0) return "";
  const finished = new Set(finishedPlayers);
  const currentIndex = players.indexOf(currentPlayerId);
  const startIndex = currentIndex < 0 ? -1 : currentIndex;
  for (let offset = 1; offset <= players.length; offset += 1) {
    const candidate = players[(startIndex + offset) % players.length];
    if (!finished.has(candidate)) return candidate;
  }
  return "";
}

export function resolveNoValidMove(
  players: readonly string[],
  currentPlayerId: string,
  finishedPlayers: readonly string[],
  diceValue: number,
): {nextPlayerId: string; keepsTurn: boolean} {
  const keepsTurn = diceValue === 6;
  return {
    nextPlayerId: keepsTurn ? currentPlayerId : nextActivePlayer(players, currentPlayerId, finishedPlayers),
    keepsTurn,
  };
}

export function finishOrderAfterMove(
  players: readonly string[],
  currentFinishOrder: readonly string[],
  movingPlayerId: string,
  movingPlayerFinished: boolean,
): string[] {
  const result: string[] = [];
  for (const playerId of currentFinishOrder) {
    if (players.includes(playerId) && !result.includes(playerId)) result.push(playerId);
  }
  if (movingPlayerFinished && !result.includes(movingPlayerId)) result.push(movingPlayerId);
  const unfinished = players.filter((id) => !result.includes(id));
  if (movingPlayerFinished && unfinished.length === 1) result.push(unfinished[0]);
  return result;
}

export function isMatchFinished(players: readonly string[], finishOrder: readonly string[]): boolean {
  return players.length >= 2 && finishOrder.length === players.length;
}

export function globalPathIndexForSeat(seatIndex: number, relativePos: number): number {
  const safeSeat = Math.max(0, Math.min(3, Math.trunc(seatIndex)));
  return (START_OFFSETS[safeSeat] - relativePos + 52) % 52;
}

function seatFor(players: readonly string[], playerSeats: Record<string, number>, playerId: string): number {
  const rawSeat = playerSeats[playerId] ?? players.indexOf(playerId);
  return Math.max(0, Math.min(3, Math.trunc(rawSeat)));
}

export function applyCaptures(
  pieces: Record<string, Piece[]>,
  playerSeats: Record<string, number>,
  players: readonly string[],
  movingPlayerId: string,
  moveDestination: MoveStep,
): CaptureResult {
  const updated = Object.fromEntries(
    Object.entries(pieces).map(([id, values]) => [id, values.map((piece) => ({...piece}))]),
  );
  if (moveDestination.inHome) return {pieces: updated, didCapture: false, capturedPieces: []};
  const movingSeat = seatFor(players, playerSeats, movingPlayerId);
  const globalDestination = globalPathIndexForSeat(movingSeat, moveDestination.pos);
  if (SAFE_GLOBAL_POSITIONS.has(globalDestination)) {
    return {pieces: updated, didCapture: false, capturedPieces: []};
  }

  const capturedPieces: MoveCapture[] = [];
  for (const opponentId of players) {
    if (opponentId === movingPlayerId) continue;
    const opponentSeat = seatFor(players, playerSeats, opponentId);
    updated[opponentId] = (updated[opponentId] ?? []).map((piece) => {
      if (piece.pos === BASE_POSITION || piece.inHome) return piece;
      if (globalPathIndexForSeat(opponentSeat, piece.pos) !== globalDestination) return piece;
      capturedPieces.push({
        playerId: opponentId,
        pieceId: piece.id,
        from: {pos: piece.pos, inHome: piece.inHome},
      });
      return {...piece, pos: BASE_POSITION, inHome: false};
    });
  }
  return {pieces: updated, didCapture: capturedPieces.length > 0, capturedPieces};
}
